crossdock/server/server.py

```diff
@@ -130,79 +166,66 @@ def handle_trace_request(self, http_request, trace_request, span_handler, respon
         if span_handler:
             span_handler(span)
 
-        trace_id = "%x" % span.trace_id
-        observed_span = ObservedSpan(
-            trace_id, span.is_sampled(),
-            span.get_baggage_item(constants.baggage_key))
+        with request_context.span_in_stack_context(span):
+            trace_id = '%x' % span.trace_id
+            observed_span = ObservedSpan(
+                trace_id, span.is_sampled(),
+                span.get_baggage_item(constants.baggage_key))
 
-        tr = TraceResponse(span=observed_span)
+            trace_response = TraceResponse(span=observed_span)
 
-        if trace_request and trace_request.downstream is not None:
-            self.call_downstream(span, trace_request, tr, response_writer)
-        else:
-            response_writer.write(serializer.obj_to_json(tr))
-            response_writer.finish()
-
-    def call_downstream(self, span, trace_request, trace_response, response_writer):
-        if trace_request.downstream.transport == Transport.HTTP:
-            self.call_downstream_http(span, trace_request, trace_response, response_writer)
-        elif trace_request.downstream.transport == Transport.TCHANNEL:
-            if tchannel_supported:
-                self.real_call_downstream_tchannel(span, trace_request, trace_response,
-                                                   response_writer)
-            else:
-                self.call_downstream_tchannel(span, trace_request, trace_response,
-                                              response_writer)
-        else:
-            response_writer.finish()
-            raise UnknownTransportException("%s" % trace_request.downstream.transport)
+            if trace_request and trace_request.downstream is not None:
 
-    def call_downstream_http(self, span, trace_request, trace_response, response_writer):
+                def handle_response(future):
+                    downstream_trace_resp = future.result()
+                    if downstream_trace_resp.notImplementedError:
+                        response_writer.write(serializer.obj_to_json(downstream_trace_resp))
+                    else:
+                        trace_response.downstream = downstream_trace_resp
+                        response_writer.write(serializer.obj_to_json(trace_response))
+                    response_writer.finish()
 
-        def handle_response(response):
-            tr = serializer.traceresponse_from_json(response.body)
-            if tr.notImplementedError:
-                response_writer.write(serializer.obj_to_json(tr))
+                future = call_downstream(span, trace_request.downstream)
+                tornado.ioloop.IOLoop.current().add_future(future, handle_response)
```

Comment:
this is not necessary, let's just have all methods be normal coroutines and you can simply call `res = yield func()` instead of coding with callbacks (which are always harder to read)